presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java

```diff
@@ -0,0 +1,672 @@
+            return true;
+        }
+        Optional<Key> normalizedKeyRequirement = keyRequirement.normalize(equivalenceClassProperty);
+        if (normalizedKeyRequirement.isPresent()) {
+            return keyProperty.satisfiesKeyRequirement(keyRequirement);
+        }
+        else {
+            return false;
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("KeyProperty", keyProperty)
+                .add("EquivalenceClassProperty", equivalenceClassProperty)
+                .add("MaxCardProperty", maxCardProperty)
+                .toString();
+    }
+
+    /**
+     * This logical properties builder should be used by PlanNode's that do not
+     * (yet perhaps) propagate or add logical properties. For example, a GroupIdNode does
+     * not propagate or add logical properties. This is the PlanNode default.
+     */
+    public static class NoPropagateBuilder
+    {
+        KeyProperty keyProperty = new KeyProperty();
+        MaxCardProperty maxCardProperty = new MaxCardProperty();
+        EquivalenceClassProperty equivalenceClassProperty = new EquivalenceClassProperty();
+
+        NoPropagateBuilder() {}
+
+        LogicalPropertiesImpl build()
+        {
+            return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
+        }
+    }
+
+    /**
+     * This logical properties builder should be used by PlanNode's that simply
+     * propagate source properties without changes. For example, a SemiJoin node
+     * propagates the inputs of its non-filtering source without adding new properties.
+     * A SortNode also propagates the logical properties of its source without change.
+     */
+    public static class PropagateBuilder
+    {
+        private final LogicalPropertiesImpl sourceProperties;
+        private final KeyProperty keyProperty = new KeyProperty();
+        private final MaxCardProperty maxCardProperty = new MaxCardProperty();
+        private final EquivalenceClassProperty equivalenceClassProperty = new EquivalenceClassProperty();
+
+        PropagateBuilder(LogicalPropertiesImpl sourceProperties)
+        {
+            requireNonNull(sourceProperties, "Properties of source must be provided.");
+            this.sourceProperties = sourceProperties;
+        }
+
+        LogicalPropertiesImpl build()
+        {
+            keyProperty.addKeys(sourceProperties.keyProperty);
+            maxCardProperty.update(sourceProperties.maxCardProperty);
+            equivalenceClassProperty.update(sourceProperties.equivalenceClassProperty);
+            return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
+        }
+    }
+
+    /**
+     * This logical properties builder is used by a TableScanNode to initialize logical properties from catalog constraints.
+     * The current implementation supports only key constraints. This will be extended
```

Comment:
Documentations talking about "future", especially "near future" are not useful. Please remove. (Saving yourself some effort when you actually implements it in the near future to seek them all out and delete them.)

Reply:
Done.